tests/test_Step.cc

```diff
@@ -627,57 +627,81 @@ TEST_CASE("execute(): Exporting variables into a context", "[Step]")
         step.execute(context);
         REQUIRE(context.variables.size() == 1);
-        REQUIRE(std::get<std::string>(context.variables["b"]) == "Test");
+        REQUIRE(std::get<LuaString>(context.variables["b"]) == "Test");
     }
 
     SECTION("Exporting an integer")
     {
         step.set_used_context_variable_names(VariableNames{ "a" });
         step.execute(context);
         REQUIRE(context.variables.size() == 1);
-        REQUIRE(std::get<long long>(context.variables["a"]) == 42);
-        REQUIRE_THROWS_AS(std::get<double>(context.variables["a"]),
+        REQUIRE(std::get<LuaInteger>(context.variables["a"]) == 42);
+        REQUIRE_THROWS_AS(std::get<LuaFloat>(context.variables["a"]),
                           std::bad_variant_access);
-        REQUIRE_THROWS_AS(std::get<std::string>(context.variables["a"]),
+        REQUIRE_THROWS_AS(std::get<LuaString>(context.variables["a"]),
+                          std::bad_variant_access);
+        REQUIRE_THROWS_AS(std::get<LuaBool>(context.variables["a"]),
                           std::bad_variant_access);
     }
 
-    SECTION("Exporting a double")
+    SECTION("Exporting a floating point")
     {
         step.set_used_context_variable_names(VariableNames{ "b" });
         step.execute(context);
         REQUIRE(context.variables.size() == 1);
 
-        REQUIRE_THROWS_AS(std::get<long long>(context.variables["b"]), std::bad_variant_access);
-        REQUIRE(std::get<double>(context.variables["b"]) == 1.5);
-        REQUIRE_THROWS_AS(std::get<std::string>(context.variables["b"]), std::bad_variant_access);
+        REQUIRE_THROWS_AS(std::get<LuaInteger>(context.variables["b"]), std::bad_variant_access);
+        REQUIRE(std::get<LuaFloat>(context.variables["b"]) == 1.5);
+        REQUIRE_THROWS_AS(std::get<LuaString>(context.variables["b"]), std::bad_variant_access);
+        REQUIRE_THROWS_AS(std::get<LuaBool>(context.variables["b"]), std::bad_variant_access);
     }
 
     SECTION("Exporting a string")
     {
         step.set_used_context_variable_names(VariableNames{ "c" });
         step.execute(context);
         REQUIRE(context.variables.size() == 1);
-        REQUIRE_THROWS_AS(std::get<long long>(context.variables["c"]), std::bad_variant_access);
-        REQUIRE_THROWS_AS(std::get<double>(context.variables["c"]), std::bad_variant_access);
-        REQUIRE(std::get<std::string>(context.variables["c"]) == "string");
+        REQUIRE_THROWS_AS(std::get<LuaInteger>(context.variables["c"]), std::bad_variant_access);
+        REQUIRE_THROWS_AS(std::get<LuaFloat>(context.variables["c"]), std::bad_variant_access);
+        REQUIRE(std::get<LuaString>(context.variables["c"]) == "string");
+        REQUIRE_THROWS_AS(std::get<LuaBool>(context.variables["c"]), std::bad_variant_access);
+    }
+
+    SECTION("Exporting a function (!?)")
+    {
+        step.set_used_context_variable_names(VariableNames{ "d" });
+        step.execute(context);
+        // silently nothing is exported
```

Comment:
I was tempted to export a `std::monostate` as placeholder for 'can not be represented'.
But that would be another PR.
Any opinions on that?
I myself only think this is like `[ ] [ ] [x] [ ] [ ]` ...

Reply:
Sounds reasonable but how will you then test for `std::monostate`? Something like `REQUIRE(context.variables[0] == std::monospace)`? This will throw because the instance `VariableTable` has no elements. Illuminate me...

Reply:
```c++
REQUIRE(std::holds_alternative<std::monostate>(context.variables["a_funtion_variable"]))
```
This is most basic variant stuff? Maybe I don't get the question.

Reply:
```diff
--- a/src/Step.cc
+++ b/src/Step.cc
@@ -69,6 +69,9 @@ void Step::copy_used_variables_from_context_to_lua(const Context& context, sol::
                 else if constexpr (std::is_same_v<T, sol::lua_nil_t>) {
                     lua[varname_str] = sol::lua_nil_t{ };
                 }
+                else if constexpr (std::is_same_v<T, std::monostate>) {
+                    lua[varname_str] = ??? :-D probably just ignore
+                }
                 else
                 {
                     static_assert(always_false_v<T>, "Unhandled type in variable import");
@@ -104,6 +107,7 @@ void Step::copy_used_variables_from_lua_to_context(const sol::state& lua, Contex
                 context.variables.erase(varname);
                 break;
             default:
+                context.variables[varname] = std::monostate{ };
                 break;
         }
     }
```

That feature allows to see on the C++ side that there is something 'unrepresentable' in the variable. It does not allow to pass it from step to step.

Reply:
Ah, when I see the complete diff...

```c++
        REQUIRE_THROWS_AS(std::get<LuaInteger>(context.variables["c"]), std::bad_variant_access);
        REQUIRE_THROWS_AS(std::get<LuaFloat>(context.variables["c"]), std::bad_variant_access);
        REQUIRE(std::get<LuaString>(context.variables["c"]) == "string");
        REQUIRE_THROWS_AS(std::get<LuaBool>(context.variables["c"]), std::bad_variant_access);
```

I do not really understand the need for all these tests, and why `std::holds_alternative` is not used (like I proposed above). This test also tests `variant`'s behavior ;) And all the exceptions... needlessly.

But to keep the style, we would

```diff
-REQUIRE(std::holds_alternative<std::monostate>(context.variables["a_funtion_variable"]));
+REQUIRE(std::get<std::monostate>(context.variables["a_funtion_variable"]) == std::monostate{ });
```

Reply:
> ```diff
> --- a/src/Step.cc
> +++ b/src/Step.cc
> ...
> +                else if constexpr (std::is_same_v<T, std::monostate>) {
> +                    lua[varname_str] = ??? :-D probably just ignore
> +                }
> ...
> @@ -104,6 +107,7 @@ void Step::copy_used_variables_from_lua_to_context(const sol::state& lua, Contex
>                context.variables.erase(varname);
>                 break;
>             default:
> +                context.variables[varname] = std::monostate{ };
>                 break;
>         }
>     }
>```
>
> That feature allows to see on the C++ side that there is something 'unrepresentable' in the variable. It does not allow to pass it from step to step.

What do you mean with the phrase _It does not allow to pass it from step to step_? I mean it is set to an empty variable (`std::monostate`) and surely it cannot be access in the successive steps. Therefore you have in the previous comment the `???`. I would remove it - so no check for `else if constexpr (std::is_same_v<T, std::monostate>)` because it has no reason or use case.

Reply:
This is just like placeholder code for explanation (to show the relevant places).
Not really intended to be implemented as such.

 >  It does not allow to pass it from step to step

Of course it _would be possible_ to pass anything from step to step, we would just need to store a `sol::object{ }` and not a `std::variant{ }`.

The problem with THAT is that the `sol::object`-object does not own the object, but it is a reference into the relevant structure that is held by some Lua virtual machine. That would prose no problem if we would reuse the same VM for every step in a sequence. It would also be more efficient, because Lua VM spin-up is rather expensive. If you reuse the VM you have to pay that price only once when the sequence is started (now: whenever a step is started).

The downside is the robustness... Maybe one step corrupted the VM... Now we start with a fresh VM at the next step.
But well, ... usually if the VM gets potentially corropted you stop the sequence anyhow.

Another downside is that being able to pass mappings or Lua classes and objects from step to step might hinder the switch to another script language (but ... that would anyhow require rewriting of all the script-near code).

And step separation becomes less. This might be good or bad.

For example with the current design you can not pass a doocs-`IIII` (or `IFFF` or 'USTR' or 'IMAGE') object from step to step, you need to pass 4 integers in 4 variables (you can not even pass a Lua integer array).

Reply:
About the `monostate`... I do not really see the usefulness. Sure, a step could find out that one of its predecessors tried to store something and failed. But what have we gained by that? This is almost the same as clearing the variable from the context.

Reply:
If I would have deemed the `monostate` stuff a must, I would have included in this PR.

But the usefullness is as such (I picture myself doing things):
* Write a script that shall interact with whatever application I write
* (This is maybe not a control system based application of the taskolib)
* Maybe the application dumps all variable values after a run into a file
* I write some ints, some floats, and an int array
* The array seems to be missing, why? (it is silently ignored)
* At first I do not notice, because my code takes all variables elements and export them
* But the lua-array variable is missing, like `nil`
* Is the transfer broken?
* If the variable had a special type 'here is a value I can not represent in C++' that would be more clear
* The export code would have err-ed out with that while doing the array

This is just, you have a greater chance to notice some value export lua->C++ is impossible. More noticeable than 'it is just ignored'.